tools/kth_threshold.cpp

```diff
@@ -0,0 +1,211 @@
+#include "util/util.hpp"
+#include "wand_data_compressed.hpp"
+#include "wand_data_raw.hpp"
+
+#include "query/algorithm.hpp"
+#include "scorer/scorer.hpp"
+
+#include "CLI/CLI.hpp"
+
+using namespace pisa;
+
+template <typename IndexType, typename WandType>
+void kt_thresholds(
+    const std::string& index_filename,
+    const std::string& wand_data_filename,
+    const std::vector<Query>& queries,
+    std::string const& type,
+    ScorerParams const& scorer_params,
+    uint64_t k,
+    bool quantized,
+    std::optional<std::string> pairs_filename,
+    std::optional<std::string> triples_filename)
+{
+    IndexType index;
+    mio::mmap_source m(index_filename.c_str());
+    mapper::map(index, m);
+
+    WandType wdata;
+
+    auto scorer = scorer::from_params(scorer_params, wdata);
+
+    mio::mmap_source md;
+    std::error_code error;
+    md.map(wand_data_filename, error);
+    if (error) {
+        spdlog::error("error mapping file: {}, exiting...", error.message());
+        std::abort();
+    }
+    mapper::map(wdata, md, mapper::map_flags::warmup);
+
+    using Pair = std::set<uint32_t>;
+    std::unordered_set<Pair, boost::hash<Pair>> pairs_set;
+
+    using Triple = std::set<uint32_t>;
+    std::unordered_set<Triple, boost::hash<Triple>> triples_set;
+
+    std::string t;
+    if (pairs_filename) {
+        std::ifstream pin(*pairs_filename);
+        while (std::getline(pin, t)) {
+            std::vector<std::string> p;
+            boost::algorithm::split(p, t, boost::is_any_of(" \t"));
+            pairs_set.insert({(uint32_t)std::stoi(p[0]), (uint32_t)std::stoi(p[1])});
+        }
+        spdlog::info("Number of pairs loaded: {}", pairs_set.size());
+    }
+
+    if (triples_filename) {
+        std::ifstream trin(*triples_filename);
+        while (std::getline(trin, t)) {
+            std::vector<std::string> p;
+            boost::algorithm::split(p, t, boost::is_any_of(" \t"));
+            triples_set.insert(
+                {(uint32_t)std::stoi(p[0]), (uint32_t)std::stoi(p[1]), (uint32_t)std::stoi(p[2])});
+        }
+        spdlog::info("Number of triples loaded: {}", triples_set.size());
+    }
+
+    for (auto const& query: queries) {
+        float threshold = 0;
+
+        auto terms = query.terms;
+        topk_queue topk(k);
+        wand_query wand_q(topk);
+
+        for(auto&& term : terms) {
+            Query q;
+            q.terms.push_back(term);
+            wand_q(make_max_scored_cursors(index, wdata, *scorer, q), index.num_docs());
+            topk.finalize();
+            auto results = topk.topk();
+            topk.clear();
+            float t = 0.0;
+            if (results.size() == k) {
+                t = results.back().first;
+            }
+            threshold = std::max(threshold, t);
+        }
+        for (size_t i = 0; i < terms.size(); ++i) {
+            for (size_t j = i + 1; j < terms.size(); ++j) {
+                if (pairs_set.count({terms[i], terms[j]})) {
```

Comment:
We should avoid implicit conversions, I'm surprised this is not rejected by clang-tidy. Let's use either `count(...) > 0` or `find(...) != pairs_set.end()`. I know it's a pain `contains` isn't in c++17 but still.

Reply:
Actually `readability-implicit-bool-conversion` is in place. Weird!